util/src/main/java/io/grpc/util/MultiChildLoadBalancer.java

```diff
@@ -398,18 +420,27 @@ protected void setDeactivated() {
       deactivated = true;
     }
 
+    protected void markReactivated() {
+      deactivated = false;
+    }
+
     protected void setResolvedAddresses(ResolvedAddresses newAddresses) {
       checkNotNull(newAddresses, "Missing address list for child");
       resolvedAddresses = newAddresses;
     }
 
+    /**
+     * The default implementation not only shuts down the lb policy, it also removes this child
+     * from the map of children maintained by the petiole policy.  Therefore, if you plan to
+     * reactivate this child later, then you must override this method in your extending policy.
```

Comment:
What is the exact scenario that we need to re-activate? Is that for the subchannels in the picker that is different from the updated picker due to handling new addresses?

The address update that removes the old addresses and deactivate childLbState would immediately remove the key from the `childLbState`. Those deactivated childLbState would never be reused in the perspective of handling an update(`acceptResolvedAddress`), because the reusing is based on the `childLbStates` map key.

I wish we do not need this override, for code readability purpose.

Reply:
I've removed the logic other than resetting the flag and left it to subclasses to do what they want with complex deactivate/reactivate pairs.  Right now, I'm using reactivate in RingHash tests and ClusterManagerLB has a delayed shutdown on deactivate.

Reply:
>  left it to subclasses to do what they want with complex deactivate/reactivate pairs. 

I see. That sounds fine.
Could you elaborate on why not keep shutdown in the default `MultiChildLB.ChildLbState.deactivate()` and ringhash use the default?

If we decide to remove shutdown from the default `deactivate()`, then you have to update other LBs, which inherit MultiChildLB, on their `ChildLBState.deactivate()`.

Reply:
The only class that overrode deactivate was ClusterManagerLoadBalancer which specifically wants the shutdown to be triggered by a timer rather than the removal from the address list.

Reply:
The only place that calls `deactivate` is in `MultiChildLB's acceptResolvedAddressesInternal,` so that method explicitly does a shutdown rather than expecting deactivate to do it. Note that logic moved to the new `MultiChildLB.updateLbStateAndShutdownRemoved` and is conditional on the protected method `shutdownInAcceptResolvedAddresses`.  Therefore, the other children don't need to be updated.